### PR TITLE
HEEDLS-495 All delegates - send single welcome email

### PR DESCRIPTION
**Pull request by ibrahimmunir14**
### JIRA link
[HEEDLS-495](https://softwiretech.atlassian.net/browse/HEEDLS-495)

### Description
When the “Send welcome email” button is clicked on “Manage delegate” page, we send the delegate a welcome email including a link to set their password. The Set Password logic and pages are heavily based on, in fact, almost identical to, Reset Password.

#### Password set page

-----
### Developer checks
I have:
- [x] Run the formatter and made sure there are no IDE errors.
- [x] Written tests for the changes (controller, data services, services, view models etc) and manually tested my work with and without JavaScript. Full manual testing guidelines can be found here: https://softwiretech.atlassian.net/wiki/spaces/HEE/pages/6703648740/Testing
- [x] Updated/added documentation in Swiki and/or Readme.
- [x] Updated my Jira ticket with information about other parts of the system that were touched as part of the MR and have to be sanity tested to ensure nothing’s broken.
- [x] Scanned over my own MR to ensure everything is as expected.

### CODE REVIEW

File: DigitalLearningSolutions.Web/Views/SetPassword/Index.cshtml [hunk 1]
```diff
@@ -0,0 +1,50 @@
+@using DigitalLearningSolutions.Web.ViewModels.Common
```
Comment (JBsoftwire) [inline DigitalLearningSolutions.Web/Views/SetPassword/Index.cshtml:1]: I think a lot of this can be commonised with the Reset Password form.

File: DigitalLearningSolutions.Web/Views/SetPassword/Index.cshtml [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+      @if (errorHasOccurred)
+      {
+        <vc:error-summary order-of-property-names="@(new[] { nameof(Model.Password), nameof(Model.ConfirmPassword) })" />
+      }
+
+      <div class="nhsuk-u-margin-bottom-8">
+        <h1 id="page-heading" class="nhsuk-heading-xl">Set password</h1>
+        <p class="nhsuk-body-l">
+          Use the form below to set the password for delegate and admin accounts associated with your email address.
```
Comment (JBsoftwire) [inline DigitalLearningSolutions.Web/Views/SetPassword/Index.cshtml:25]: This isn't in scope for this review so doesn't need to be addressed as part of it, but are we making a system where someone (admin or delegate) making a new delegate account at a new centre is asked to reset their established password? That will be odd UX.

File: DigitalLearningSolutions.Data/Services/PasswordResetService.cs [hunk 3]
```diff
@@ -82,7 +83,7 @@ public async Task InvalidateResetPasswordForEmailAsync(string email)
             }
         }
 
-        private string GenerateResetPasswordHash(User user)
+        public string GenerateResetPasswordHash(User user)
```
Comment (JBsoftwire) [inline DigitalLearningSolutions.Data/Services/PasswordResetService.cs:86]: It's a bit organisationally odd to have the reset email generated and dispatched from inside the password-handling service, but the set email generated and dispatched from outside. This is more of a matter of preference though so if you prefer it this way feel free to keep it.

**Review by JBsoftwire — CHANGES_REQUESTED** [review]

**Review by JBsoftwire — APPROVED** [review]

File: DigitalLearningSolutions.Web/Views/SetPassword/Error.cshtml [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+}
+
+@section NavMenuItems {
+  <partial name="_NavMenuItems" />
+}
+
+<h1 class="nhsuk-heading-xl">Something went wrong...</h1>
+<p class="nhsuk-body-m">
+  There was a problem resetting your password.
+  The set password link is only valid once, and expires after 2 hours.
```
Comment (stellake) [inline DigitalLearningSolutions.Web/Views/SetPassword/Error.cshtml:12]: I added this to the ticket - can we make this link expire in three days instead of 2hours please (only for setting the pw, resetting should stay as it was)

File: DigitalLearningSolutions.Web/Views/SetPassword/Error.cshtml [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+  ViewData["Title"] = "Set Password: Error";
+}
+
+@section NavMenuItems {
+  <partial name="_NavMenuItems" />
+}
+
+<h1 class="nhsuk-heading-xl">Something went wrong...</h1>
+<p class="nhsuk-body-m">
+  There was a problem resetting your password.
```
Comment (stellake) [inline DigitalLearningSolutions.Web/Views/SetPassword/Error.cshtml:11]: setting?

File: DigitalLearningSolutions.Web/Controllers/TrackingSystem/Delegates/ViewDelegateController.cs [hunk 6]
```diff
@@ -48,5 +53,32 @@ public IActionResult Index(int delegateId)
...
+            if (delegateUser == null || delegateUser.CentreId != centreId)
+            {
+                return new NotFoundResult();
+            }
+
+            registrationService.GenerateAndSendDelegateWelcomeEmail(delegateUser!);
+
+            TempData.Set(new WelcomeEmailSentViewModel(delegateUser));
+
+            return RedirectToAction("WelcomeEmailSent", new { delegateId = delegateUser.Id });
```
Comment (stellake) [inline DigitalLearningSolutions.Web/Controllers/TrackingSystem/Delegates/ViewDelegateController.cs:72]: Could we just show the successful View here instead of dealing with TempData & redirects?

File: DigitalLearningSolutions.Data/Services/PasswordResetService.cs [hunk 7]
```diff
@@ -19,6 +19,7 @@ public interface IPasswordResetService
         Task<bool> EmailAndResetPasswordHashAreValidAsync(string emailAddress, string resetHash);
         void GenerateAndSendPasswordResetLink(string emailAddress, string baseUrl);
         Task InvalidateResetPasswordForEmailAsync(string email);
+        public string GenerateResetPasswordHash(User user);
```
Comment (stellake) [inline DigitalLearningSolutions.Data/Services/PasswordResetService.cs:22]: Do we need the "public" here as everything in interface should be public anyway?

File: DigitalLearningSolutions.Data/Services/RegistrationService.cs [hunk 8]
```diff
@@ -169,5 +193,38 @@ bool refactoredTrackingSystemEnabled
...
+        {
+            const string emailSubject = "Welcome to Digital Learning Solutions - Verify your Registration";
+            var setPasswordUrl = $"{config["AppRootPath"]}/SetPassword?code={setPasswordHash}&email={emailAddress}";
+            
+            BodyBuilder body = new BodyBuilder
+            {
+                TextBody = $@"Dear {firstName} {lastName},
+                            An administrator has registered your details to give you access to the Digital Learning Solutions (DLS) platform under the centre {centreName}.
+                            You have been assigned the unique DLS delegate number {candidateNumber}.
+                            To complete your registration and access your Digital Learning Solutions content, please follow this link:{setPasswordUrl}
```
Comment (stellake) [inline DigitalLearningSolutions.Data/Services/RegistrationService.cs:214]: Could we have a space after ":"?

File: DigitalLearningSolutions.Data/Services/RegistrationService.cs [hunk 9]
```diff
@@ -169,5 +193,38 @@ bool refactoredTrackingSystemEnabled
...
+                            Note that this link can only be used once.
+                            Please don't reply to this email as it has been automatically generated.",
+                HtmlBody = $@"<body style= 'font - family: Calibri; font - size: small;'>
+                                <p>Dear {firstName},</p>
+                                <p>An administrator has registered your details to give you access to the Digital Learning Solutions (DLS) platform under the centre {centreName}.</p>
+                                <p>You have been assigned the unique DLS delegate number {candidateNumber}.</p>
+                                <p>To complete your registration and access your Digital Learning Solutions content, please follow this link: <a href=""{setPasswordUrl}"">{setPasswordUrl}</a></p>
```
Comment (stellake) [inline DigitalLearningSolutions.Data/Services/RegistrationService.cs:221]: How does this hyperlink read out with a screen reader? Would it be better to name it something more clear like "Click here to complete your registration..."

We have a little bit of wiggle room to change things here to be more accessible :)

File: DigitalLearningSolutions.Data/Services/RegistrationService.cs [hunk 10]
```diff
@@ -169,5 +193,38 @@ bool refactoredTrackingSystemEnabled
...
+                                <p>Note that this link can only be used once.</p>
```
Comment (stellake) [inline DigitalLearningSolutions.Data/Services/RegistrationService.cs:222]: Can we also add a sentence about the link expiring in 3 days.

File: DigitalLearningSolutions.Data/Services/PasswordResetService.cs [hunk 11]
```diff
@@ -82,7 +83,7 @@ public async Task InvalidateResetPasswordForEmailAsync(string email)
             }
         }
 
-        private string GenerateResetPasswordHash(User user)
+        public string GenerateResetPasswordHash(User user)
         {
             string hash = Guid.NewGuid().ToString();
```
Comment (stellake) [inline DigitalLearningSolutions.Data/Services/PasswordResetService.cs:89]: I think the email generation file structure should be consistent between Reset & Set password so worth shuffling things around

File: DigitalLearningSolutions.Data/Services/RegistrationService.cs [hunk 12]
```diff
@@ -169,5 +193,38 @@ bool refactoredTrackingSystemEnabled
...
+            
+            BodyBuilder body = new BodyBuilder
+            {
+                TextBody = $@"Dear {firstName} {lastName},
+                            An administrator has registered your details to give you access to the Digital Learning Solutions (DLS) platform under the centre {centreName}.
+                            You have been assigned the unique DLS delegate number {candidateNumber}.
+                            To complete your registration and access your Digital Learning Solutions content, please follow this link:{setPasswordUrl}
+                            Note that this link can only be used once.
+                            Please don't reply to this email as it has been automatically generated.",
+                HtmlBody = $@"<body style= 'font - family: Calibri; font - size: small;'>
```
Comment (stellake) [inline DigitalLearningSolutions.Data/Services/RegistrationService.cs:217]: Do we need spaces between font-family and font-size (the hyphens)?

File: DigitalLearningSolutions.Data.Tests/Services/RegistrationServiceTests.cs [hunk 13]
```diff
@@ -53,6 +53,7 @@ public class RegistrationServiceTests
         private ICentresDataService centresDataService = null!;
```
Comment (stellake) [inline DigitalLearningSolutions.Data.Tests/Services/RegistrationServiceTests.cs:53]: Could we have some tests for the new functionality too?

File: DigitalLearningSolutions.Web/Views/SetPassword/Index.cshtml (hunk 2 above)
Comment (stellake) [reply DigitalLearningSolutions.Web/Views/SetPassword/Index.cshtml:25]: This functionality is used when a new delegate is created by admins, but no password is set for the account. So delegate won't need to register as the centre does it for them, but will only need to set pw. Hope this makes sense.

File: DigitalLearningSolutions.Data/Services/PasswordResetService.cs [hunk 14]
```diff
@@ -82,6 +97,27 @@ public async Task InvalidateResetPasswordForEmailAsync(string email)
...
+
+            string setPasswordHash = GenerateResetPasswordHash(delegateUser);
+            var welcomeEmail = GenerateWelcomeEmail(
+                emailAddress,
+                setPasswordHash,
+                baseUrl,
+                delegateUser.FirstName,
+                delegateUser.LastName,
+                delegateUser.CentreName,
+                delegateUser.CandidateNumber
```
Comment (stellake) [inline DigitalLearningSolutions.Data/Services/PasswordResetService.cs:116]: Is there a reason we're not passing in delegateUser directly?

File: DigitalLearningSolutions.Data/Services/RegistrationService.cs [hunk 15]
```diff
@@ -25,21 +25,24 @@ bool refactoredTrackingSystemEnabled
...
             IPasswordDataService passwordDataService,
+            IPasswordResetService passwordResetService,
             IEmailService emailService,
             ICentresDataService centresDataService,
             IConfiguration config
         )
         {
             this.registrationDataService = registrationDataService;
             this.passwordDataService = passwordDataService;
+            this.passwordResetService = passwordResetService;
```
Comment (stellake) [inline DigitalLearningSolutions.Data/Services/RegistrationService.cs:45]: Are these changes needed anymore?

File: DigitalLearningSolutions.Web/Controllers/SetPasswordController.cs [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+            {
+                return View(viewModel);
+            }
+
+            await ChangePassword(viewModel, resetPasswordData);
+
+            return View("Success");
+        }
+
+
```
Comment (stellake) [inline DigitalLearningSolutions.Web/Controllers/SetPasswordController.cs:65]: A lot of spaces :)

File: DigitalLearningSolutions.Web/Controllers/SetPasswordControllerBase.cs [hunk 17]
```diff
@@ -2,81 +2,55 @@ namespace DigitalLearningSolutions.Web.Controllers
...
     using System.Threading.Tasks;
     using DigitalLearningSolutions.Data.Services;
-    using DigitalLearningSolutions.Web.Extensions;
     using DigitalLearningSolutions.Web.Models;
-    using DigitalLearningSolutions.Web.ServiceFilter;
     using DigitalLearningSolutions.Web.ViewModels.Common;
     using Microsoft.AspNetCore.Mvc;
 
-    public class SetPasswordControllerBase : Controller
+    public abstract class SetPasswordControllerBase : Controller
```
[the remark this answers is not in the document]
Comment (AlexJacksonDS) [reply DigitalLearningSolutions.Web/Controllers/SetPasswordControllerBase.cs:9]: TBH I don't think we should commonise stuff with base controllers ever, particularly actions as they were before I started doing markups for this ticket as this means that it breaks the Visual Studio searching for views. If it's something we need in multiple controllers in a sensible way, it should really be in a helper or service. 
I don't think it's really worth setting up something to commonise these further at this point.

File: DigitalLearningSolutions.Web/Helpers/ConfigHelper.cs [hunk 18]
```diff
@@ -1,6 +1,7 @@
 ﻿namespace DigitalLearningSolutions.Web.Helpers
 {
     using System;
+    using System.Drawing;
```
Comment (stellake) [inline DigitalLearningSolutions.Web/Helpers/ConfigHelper.cs:4]: Is this needed?

File: DigitalLearningSolutions.Data.Tests/Helpers/ResetPasswordHelperTests.cs [hunk 19]
```diff
@@ -36,7 +36,37 @@ public void Reset_Password_Is_Invalid_121_Minutes_After_Creation()
...
+        [Test]
+        public void Set_Password_Is_Valid_4319_Minutes_After_Creation()
+        {
+            // Given
+            var createTime = DateTime.UtcNow;
+            var resetPassword = Builder<ResetPassword>.CreateNew()
+                .With(rp => rp.PasswordResetDateTime = createTime).Build();
+
+            // When
+            var resetIsValid = resetPassword.IsStillValidAt(createTime + TimeSpan.FromMinutes(4319), ResetPasswordHelpers.SetPasswordHasExpiryTime);
```
Comment (stellake) [inline DigitalLearningSolutions.Data.Tests/Helpers/ResetPasswordHelperTests.cs:54]: "HasExpiryTime" is missing a "h"

**Review by stellake — APPROVED** [review]
Looks good - just two minor comments about unused imports and a typo. No need for rereview though.

File: DigitalLearningSolutions.Web/Helpers/ConfigHelper.cs (hunk 18 above)
Comment (AlexJacksonDS) [reply DigitalLearningSolutions.Web/Helpers/ConfigHelper.cs]: I keep getting these caused by resharper thinking it knows better, in this case it was Icon instead of IConfiguration i think, and the cleanup code doesn't reliably remove unused usings.
